src/ApiCore/ApiTransportTrait.php

```diff
@@ -0,0 +1,144 @@
+use Google\Auth\HttpHandler\HttpHandlerFactory;
+
+trait ApiTransportTrait
+{
+    use ArrayTrait;
+
+    private $agentHeaderDescriptor;
+
+    private function setCommonDefaults(array $options)
+    {
+        $options += [
+            'enableCaching' => true,
+            'authCache' => new MemoryCacheItemPool(),
+            'authCacheOptions' => null,
+            'authHttpHandler' => HttpHandlerFactory::build(),
+            'libName' => null,
+            'libVersion' => null,
+            'gapicVersion' => null,
+        ];
+
+        if (empty($options['credentialsLoader'])) {
+            $this->validateNotNull($options, ['scopes']);
+            $options['credentialsLoader'] = $this->getADCCredentials(
+                $options['scopes'],
+                $options['authHttpHandler']
+            );
+        }
+
+        if ($options['enableCaching']) {
+            $options['credentialsLoader'] = new FetchAuthTokenCache(
+                $options['credentialsLoader'],
+                $options['authCacheOptions'],
+                $options['authCache']
+            );
+        }
+
+        $this->agentHeaderDescriptor = new AgentHeaderDescriptor([
+            'libName' => $options['libName'],
+            'libVersion' => $options['libVersion'],
+            'gapicVersion' => $options['libVersion']
+        ]);
+
+        return $options;
+    }
+
+    /**
+     * @param Call $call
+     * @param CallSettings $settings The call settings to use for this call.
+     *
+     * @return PromiseInterface
+     */
+    public function startCall(Call $call, CallSettings $settings)
+    {
+        $callable = $this->getCallable($settings);
+        return $callable($call, $settings);
+    }
+
+    /**
+     * @param Call $call
+     * @param CallSettings $settings The call settings to use for this call.
+     * @param PageStreamingDescriptor $descriptor
+     *
+     * @return PromiseInterface
+     */
+    public function getPagedListResponse(Call $call, CallSettings $settings, PageStreamingDescriptor $descriptor)
+    {
+        return new PagedListResponse(
+            $call,
+            $settings,
+            $this->getCallable($settings),
+            $descriptor
+        );
+    }
+
+    /**
+     * @param callable $callable A callable to make the API call through.
+     * @param CallSettings $settings The call settings to use for this call.
+     *
+     * @return callable
+     */
+    private function createCallStack(
+        callable $callable,
+        CallSettings $settings
+    ) {
+        $callable = new Middleware\AgentHeaderMiddleware($callable, $this->agentHeaderDescriptor);
+
+        $retrySettings = $settings->getRetrySettings();
+        if ($retrySettings && $retrySettings->retriesEnabled()) {
+            $callable = new Middleware\RetryMiddleware($callable);
+        }
+
+        return $callable;
+    }
+
+    /**
+     * Gets credentials from ADC. This exists to allow overriding in unit tests.
```

Comment:
Is this comment still true? If not - can this method be removed completely?

Reply:
It should still hold true, although the current state of the tests may not currently reflect it.